src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+
+    /** Contains symbols of fields and methods that have warnings suppressed.
+     */
+    private final Set<Symbol> suppressed = new HashSet<>();
+
+    /** The declaring class of the constructor we're currently analyzing.
+     *  This is the 'this' type we're trying to detect leaks of.
+     */
+    private JCClassDecl targetClass;
+
+    /** Snapshots of {@link #callStack} where possible 'this' escapes occur.
+     */
+    private ArrayList<DiagnosticPosition[]> warningList = new ArrayList<>();
+
+// These fields are scoped to the CONSTRUCTOR BEING ANALYZED
+
+    /** The declaring class of the "invoked" method we're currently analyzing.
+     *  This is either the analyzed constructor or some method it invokes.
+     */
+    private JCClassDecl methodClass;
+
+    /** The current "call stack" during our analysis. The first entry is some method
+     *  invoked from the target constructor; if empty, we're still in the constructor.
+     */
+    private final ArrayDeque<DiagnosticPosition> callStack = new ArrayDeque<>();
+
+    /** Used to terminate recursion in {@link #invokeInvokable invokeInvokable()}.
+     */
+    private final Set<Pair<JCTree, RefSet<Ref>>> invocations = new HashSet<>();
+
+    /** Snapshot of {@link #callStack} where a possible 'this' escape occurs.
+     *  If non-null, a 'this' escape warning has been found in the current
+     *  constructor statement, initialization block statement, or field initializer.
+     */
+    private DiagnosticPosition[] pendingWarning;
+
+// These fields are scoped to the CONSTRUCTOR OR INVOKED METHOD BEING ANALYZED
+
+    /** Current lexical scope depth in the constructor or method we're currently analyzing.
+     *  Depth zero is the outermost scope. Depth -1 means we're not analyzing.
+     */
+    private int depth = -1;
+
+    /** Possible 'this' references in the constructor or method we're currently analyzing.
+     *  Null value means we're not analyzing.
+     */
+    private RefSet<Ref> refs;
+
+// Constructor
+
+    ThisEscapeAnalyzer(Names names, Types types, Log log, Lint lint) {
+        this.names = names;
+        this.types = types;
+        this.log = log;
+        this.lint = lint;
+    }
+
+//
+// Main method
+//
+
+    public void analyzeTree(Env<AttrContext> env) {
+        this.analyzeTree(env, env.tree);
+    }
+
+    public void analyzeTree(Env<AttrContext> env, JCTree tree) {
+
+        // Sanity check
+        Assert.check(this.checkInvariants(false, false));
+
+        // Short circuit if warnings are totally disabled
+        if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+            return;
+
+        // Build a mapping from symbols of methods to their declarations.
+        // Classify all ctors and methods as analyzable and/or invokable.
+        // Track which methods and variables have warnings are suppressed.
+        new TreeScanner() {
+
+            private Lint lint = ThisEscapeAnalyzer.this.lint;
```

Comment:
Two things...

We need to support annotations on field declarations because their initializers are effectively mini-constructors. But we don't need it on local variables, parameters, etc. - too confusing.

For annotations on methods, yes it's debatable. It can be handy if you have e.g. an `init()` method that all of your constructors invoke. However, I agree it's also confusing, so I will remove.

But we should keep the notion that if a constructor invokes `this()`, and the invoked constructor has annotations suppressed, then we skip over the constructor invocation.

I will make these updates.

Reply:
Yep - I guess there's a general theme of "where do we want the warnings to be reported". My general feeling is that reporting a warning on the constructor might be enough - but I see that you try to generate a warning in the exact spot where the leakage happens - which I'm not sure if it's ok, or too clever.